Extensions: Wrap extension components with a sandbox wrapper div

leventebalogh opened the pull request:
### What changed?
This PR wraps extension components with a div that the sandbox finds while traversing up the DOM tree, and allows the plugins to access any DOM element objects within the wrapper div (e.g. target elements in an onClick even handler).

- [x] Wrap extension components with a sandbox div
- [x] Add tests
- [x] Test out manually if the wrapper div works for
  - Added components
  - Added links -> modals
  - Exposed components

academo commented on public/app/features/plugins/extensions/utils.tsx:
```diff
@@ -73,6 +75,7 @@ export const wrapWithPluginContext = <T,>(pluginId: string, Component: React.Com
       loading,
       value: pluginMeta,
     } = useAsync(() => getPluginSettings(pluginId, { showErrorAlert: false }));
+    const { value: isSandboxEnabled } = useAsync(() => isPluginFrontendSandboxEligible({ isAngular: false, pluginId }));
```
you should be using `shouldLoadPluginInFrontendSandbox` instead

academo commented on public/app/features/plugins/extensions/utils.tsx:
```diff
@@ -91,6 +94,16 @@ export const wrapWithPluginContext = <T,>(pluginId: string, Component: React.Com
+    if (isSandboxEnabled) {
+      return (
+        <PluginContextProvider meta={pluginMeta}>
+          <div data-plugin-sandbox={pluginMeta.id}>
```
my advice is to always return the wrapper div and only add the `data-plugin-sandbox` when the sandbox is on.

We might introduce side effects if we add an extra div based on conditions, better to always have the extra div.

leventebalogh replied:
Good point, updated 👍

leventebalogh committed:
feat(sandbox): wrap extension components with sandbox wrapper div

leventebalogh committed:
fix: remove circular dependency

This was caused by sandbox_plugin_loader_registry -> helpers -> backend_srv -> ... plugin_loader -> setup.

leventebalogh committed:
fix: add dependency to react hook

leventebalogh committed:
tests: add tests for extensions sandbox wrapper

leventebalogh committed:
fix: only wrap modal content after plugin loading

leventebalogh committed:
chore: remove unused code

leventebalogh commented:
/deploy-to-hg

leventebalogh committed:
Wip

hugohaggmark reviewed (commented):
Did a quick pass, I need to test this further

leventebalogh committed:
Merge branch 'main' into lb/extensions-sandbox-wrapping

* main: (294 commits)
  Update dependency react-draggable to v4.5.0 (#107647)
  Auth: make `skipTokenRotationIfRecent` feature flag enabled by default (#107676)
  Dashboards: Fix persisting repeated panels resize after edit (#107386)
  Chore: reduces barrel files (#107512)
  Alerting: Remove ruler from alert list view2 (#106778)
  datasources: remove empty if (#105314)
  I18n: Download translations from Crowdin (#107659)
  Auth: fix role_attribute_path when using ID tokens in Gitlab OAuth (#107634)
  Loki: Add scope filters to queries on the frontend (#104638)
  Loki: Fix label browser not sorted after selection of a label (#107394)
  Update dependency msw to v2.10.3 (#107643)
  Doc: Remove dynamic dashboards section (#107642)
  Update dependency marked-mangle to v1.1.11 (#107640)
  Extensions: Undo removal of required enterprise dependencies (#107641)
  feat(config): add config `ensure_default_org_and_user` (#107619)
  Update typescript-eslint monorepo to v8.35.1 (#107633)
  Eslint: Allow 'unset' in no-border-radius-literal lint rule (#106619)
  Update dependency @prometheus-io/lezer-promql to v0.304.2 (#107631)
  fix empty folder form on root folder level (#107630)
  Internationalisation: Enable automatic scheduling for translations (#107629)
  ...

leventebalogh committed:
review(actions.ts): extract logic to function

leventebalogh commented on public/app/features/plugins/admin/state/actions.ts:
```diff
@@ -202,7 +203,11 @@ export const install = createAsyncThunk<
+      .get('/api/frontend/settings')
+      .then((settings: Settings) => {
+        config.panels = settings.panels;
+      });
```
Fixed in https://github.com/grafana/grafana/pull/103064/commits/92d2d3e1052b1c62dbdadc90e7ff5bb8cc9e54e4.

leventebalogh commented:
/deploy-to-hg

leventebalogh committed:
test: remove testing app id

leventebalogh commented on public/app/features/plugins/extensions/utils.tsx:
```diff
@@ -102,13 +104,18 @@ const getModalWrapper = ({
     return (
       <Modal title={title} className={className} isOpen onDismiss={onDismiss} onClickBackdrop={onDismiss}>
-        <Body onDismiss={onDismiss} />
+        <div {...(isSandboxEnabled && { 'data-plugin-sandbox': pluginId })} data-testid="plugin-sandbox-wrapper">
```
@academo I was wondering if we should be waiting for the `shouldLoadPluginInFrontendSandbox()` to load before rendering anything.

(It could be risky if the sandbox availability check would hang, because then the UI will not render. I think if the sandbox is enabled and we render something before the `shouldLoadPluginInFrontendSandbox()` check would complete, then the worst that can happen is that an extension doesn't have access to elements inside that div for a short timewindow.)

leventebalogh committed:
Merge branch 'main' into lb/extensions-sandbox-wrapping

* main:
  Plugin Loader: Don't load an app plugin multiple times (#107579)
  unistore: add more kvstore tests (#107692)
  Alerting: Add support for "alerting.notifications.routes.conflictingMatchers" (#107687)
  Update Advisor docs (#107683)
  Plugin Extensions: Add error boundaries (#107515)
  chore(datasources): Adding debug logging to datasource provisioning when resource is not found in SqlStore (#107652)
  Dashboard: Fix cache validation to prevent stale cache (#105918)
  Prometheus: display a warning when subquery is used (#106199)
  Postgres: Regroup config page for TLS related settings (#107686)
  Internationalisation: Mark up `geomap` panel edit options (#107140)
  unistore: add custom kv grpc (#107464)
  Update dependency swagger-ui-react to v5.26.1 (#107681)
  Update dependency react-i18next to v15.6.0 (#107680)
  Update dependency react-hook-form to v7.60.0 (#107677)

leventebalogh committed:
tests: fix extension utils tests

leventebalogh committed:
tests: remove unnecessary code

jackw reviewed (approved):
LGTM! 🚀

leventebalogh commented on public/app/features/plugins/extensions/utils.tsx:
```diff
@@ -133,7 +133,9 @@ const getModalWrapper = ({
           log={baseLog}
         >
-          <Body onDismiss={onDismiss} />
+          <div data-plugin-sandbox={config.pluginId} data-testid="plugin-sandbox-wrapper">
```
I think that's a really good point mate, I was thinking about the same yesterday. 

Yesterday we had a chat with Esteban, and we decided to do it like this (attribute always there) because:
- the "is-sandbox-enabled-for-plugin" check is an async one, and if we wait for it, which can introduce an extra delay to the rendering of the extension
- the check is introducing an extra complexity to the extension rendering logic
-  the sandbox wrapper div is is not doing anything if the sandbox is disabled

I'll create an issue today about how we could easily see if a plugin is sandboxed.
